### ISSUE TITLE
I would really like to implement an INavigableMap but I need some help to get started

### ISSUE THREAD

**Issue by KIC** [hazelcast/hazelcast#5419]
I need to calculate things like Moving Averages or Curve Fittings on a very big set of data. Hazelcast fits very well the needs of clustering and distributing the work to the data (-node). But as you can imagine for such calculations you need ordered data. Today I am simply storing my data into an IMap[Interger, TreeMap]. This way I have some kind of partitioning because my key is ordered and the value as well so I can distribute calculations to ordered data on a cluster.

But there are some drawbacks like (de-)serializing a whole TreeMap for a slice of data, so I would really like to implement an INavigableMap. But the current code base is very huge and difficult to debug/reverse engineer. For example with plain debugging you are not able to find the real storage of a value in a map. All the proxy classes make this very difficult - so I would really like to get some help in the sens of: 
1) What is the process for a map create, put and get?
2) How would you implement a new map?
3) How can I get started, which classes, test cases should I explore first?

These are my thoughts for the implementation. Maybe you could point out some issues I have to expect too.

Features:
The map must be sorted (ConcurentSkipListMap as RecordStore)
The map must provide a subMap method like #sumMap(fromKey, toKey, forceReload)
The NavigableMapStore interface provides a load(fromKey, toKey) method. In the case of someone calls subMap(from, to) the MapStore#load should load the delta from the last/first key in the map if forceReload is false. if forceRelaod is true we reload all the keys in the specified range.
stats, event, monitors should be similar to IMap

Pitfalls:

Partitioning:
We have to make sure that the partitions are also ordered in a way otherwise we had to reorder the keys erverytime. So the user have to implement in the Key objects the Partitionaware. Instead of using the Hash, the partition awareness gives an ordered integer i.e. der year of a date.

Backup:
The backup strategy should not mess up existing ordered maps in the partition. So for backup we need a seperate partition

Combining of keys over partitions:
When a user requests a key range over multiple partitions the combined map should be like a LinkedHashMap instead of a TreeMap because we already have the ordering by ordered partition keys and ordered maps in each partition.

Thanks for your help!
KIC

**Comment by Tembrel** [comment]
The glory of Hazelcast is `IMap`, a convincing illusion of always-available more-or-less uniform access to distributed associative data. It's close enough to the behavior of a `Map` that `IMap` extends `Map`, with some important caveats about how it violates the `Map` contract. It's still an illusion, though, with Hazelcast managing the nasty reality of distributed data behind the curtain.

The success of this illusion makes it very tempting to expect that an apparently similar type like `NavigableMap` could also be distributed efficiently. But the forces that are balanced and the rules that are bent to bring `IMap` as close as it is to being a real `Map` are also the forces that make it very hard to impose further restrictions on the data structure.

In particular, the use of partitions, an essential aspect of the `IMap` implementation, mitigates against ordering of keys. If the partition tries to respect the ordering, then you give up on the balance between members of a partition, leading to poor performance. If the partition does __not__ respect the ordering, iteration in order will have very poor locality of reference, bouncing between nodes of a cluster willy-nilly and again leading to poor performance. 

In other words, there will be serious trade-offs depending on your intended use of the data structure. Rather than try to build a general `NavigableMap` on top of Hazelcast, I suggest that building a custom solution on top of `IMap` that addresses only your __immediate__ requirements and patterns of access. Share that solution with the Hazelcast community, by all means.

**Comment by noctarius** [comment]
I would agree with @Tembrel  but it might still be possible by splitting the data into a lookup table which handles ordering (but itself is partitioned by chunking - like every 100 keys use a different partition and link those by reference-keys) and storing the values completely partitioned but it might be a bigger chunk of work.

Maybe someone else has an awesome idea, storing data distributed is mostly about efficient ideas ;)

**Comment by Tembrel** [comment]
Oh, I'm definitely not saying give up, just that the first thing you do should __not__ be to define a general-purpose type `INavigableMap`. First thing should be to address __your__ immediate needs with a custom-built abstraction implemented in terms of the public Hazelcast API.

For example, your idea of "every 100 keys use a different partition" suggests that you're dealing with a known distribution of keys. That's something you can take advantage of that a general-purpose implementation could not. Every step away from the general case that you can take will be worth it in terms of increased opportunity for efficiency.

**Comment by noctarius** [comment]
Well you don't need to have a known set of keys, you just need to know the first partition and the first key, than you can kind of chain the partitions but yeah sure a more non-general solution might be easier, totally agree.

**Comment by KIC** [comment]
This is [how I do it today](https://github.com/KIC/distributedNavigableMap). The problem with this solution is, that I have to lock the sorted partition data for any write action even if I just update or add one key/value in the ConcurrentSkipListMap. And I always have to "put"-back the SkipListMap to the IMap which causes serialization (for IMap backups) of the whole SkipListMap just because one element changed in the sorted map. To avoid this issue I have tried to use a mapDB as off heap sorted map. Because then I only have a reference of an OffHeapMap inside of the IMap. Therefore there is need to lock the whole map for single key updates. But backup-synchronization is a big issue then.

So a more general solution would become more and more preferable.

**Comment by mdogan** [comment]
Closing since this is not an issue report. 

Please use google group: https://groups.google.com/forum/#!forum/hazelcast
